xmake/rules/c++/modules/modules_support/clang/builder.lua

```diff
@@ -0,0 +1,343 @@
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+--
+-- Copyright (C) 2015-present, TBOOX Open Source Group.
+--
+-- @author      ruki, Arthapz
+-- @file        clang/builder.lua
+--
+
+-- imports
+import("core.base.json")
+import("core.base.option")
+import("core.base.semver")
+import("utils.progress")
+import("private.action.build.object", {alias = "objectbuilder"})
+import("core.tool.compiler")
+import("core.project.config")
+import("core.project.depend")
+import("compiler_support")
+import(".builder", {inherit = true})
+
+-- get flags for building a module
+function _make_modulebuildflags(target, provide, bmifile, opt)
+
+    -- get flags
+    local module_outputflag = compiler_support.get_moduleoutputflag(target)
+
+    local flags
+    if module_outputflag and provide and not opt.external then -- one step compilation of named module, clang >= 16
+        flags = {{"-x", "c++-module", module_outputflag .. bmifile, "-o", target:objectfile(opt.sourcefile), "-c", opt.sourcefile}}
+    elseif provide then -- two step compilation of named module
+        flags = {{"-x", "c++-module", "--precompile", "-o", bmifile, "-c", opt.sourcefile}}
+
+        if not opt.external then
+           table.insert(flags, {"-o", target:objectfile(opt.sourcefile), "-c", opt.sourcefile})
+        end
+    else -- internal module, no bmi needed
+        flags = {{"-x", "c++", "-o", target:objectfile(opt.sourcefile), "-c", opt.sourcefile}}
+    end
+
+    if opt.name == "std" or opt.name == "std.compat" then
+       table.join2(flags[1], {"-Wno-include-angled-in-module-purview", "-Wno-reserved-module-identifier"})
+       if flags[2] then
+           table.join2(flags[2], {"-Wno-include-angled-in-module-purview", "-Wno-reserved-module-identifier"})
+       end
+    end
+
+    return table.unpack(flags)
+end
+
+-- get flags for building a headerunit
+function _make_headerunitflags(target, headerunit, bmifile)
+
+    local module_headerflag = compiler_support.get_moduleheaderflag(target)
+    assert(module_headerflag, "compiler(clang): does not support c++ header units!")
+
+    local local_directory = (headerunit.type == ":quote") and {"-I" .. path.directory(headerunit.path)} or {}
+
+    local headertype = (headerunit.type == ":angle") and "system" or "user"
+
+    local flags = table.join(local_directory, {"-xc++-header", "-Wno-everything", module_headerflag .. headertype, "-o", bmifile, "-c", headerunit.path})
+
+    return flags
+end
+
+-- do compile
+function _compile(target, flags, cppfile)
+
+    local dryrun = option.get("dry-run")
+    local compinst = target:compiler("cxx")
+    local compflags = compinst:compflags({sourcefile = cppfile, target = target})
+    local msvc = target:toolchain("msvc")
+    local flags = table.join(compflags or {}, flags)
+
+    -- trace
+    vprint(compinst:program(), table.unpack(flags))
```

Comment:
please use `os.vrunv`

Reply:
and why not use `compinst:compile` and `compinst:compcmd`?

https://github.com/xmake-io/xmake/blob/1d4a3e7645143f98ea1089b63153834f7c913d4b/xmake/rules/c%2B%2B/modules/modules_support/clang.lua#L302

Reply:
https://github.com/xmake-io/xmake/blob/1d4a3e7645143f98ea1089b63153834f7c913d4b/xmake/rules/c%2B%2B/modules/modules_support/clang.lua#L293

Reply:
because now, i'm build external modules without building the objectfile, and compcmd / compinst interfaces need an objectfile as parameter

Reply:
I don't understand, I see the sourcefile parameter set, why is there no objectfile ?

Reply:
> I don't understand, I see the sourcefile paramete, why is there no objectfile ?

when you build a module you generate a bmi file and an objectfile, if the target is a library the object file is embed inside the library so executable targets only need the bmi file so to speed up the compilation i only build the bmi file without the useless objectfile when the module came from a library target (or a package)

Reply:
and headerunits doesn't generate objectfile